WalletWasabi/WabiSabi/Client/RoundStatusUpdater.cs

```diff
@@ -0,0 +1,96 @@
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Bases;
+using WalletWasabi.WabiSabi.Backend.PostRequests;
+using WalletWasabi.WabiSabi.Backend.Rounds;
+using WalletWasabi.WabiSabi.Models;
+
+namespace WalletWasabi.WabiSabi.Client
+{
+	public class RoundStatusUpdater : PeriodicRunner
+	{
+		private IWabiSabiApiRequestHandler ArenaRequestHandler { get; }
+		private ImmutableDictionary<uint256, RoundState> RoundStates { get; set; } = ImmutableDictionary.Create<uint256, RoundState>();
+
+		private Dictionary<uint256, List<(TaskCompletionSource<RoundState> Task, Predicate<RoundState> Predicate)>> Awaiters { get; set; } = new();
+
+		public RoundStatusUpdater(TimeSpan requestInterval, IWabiSabiApiRequestHandler arenaRequestHandler) : base(requestInterval)
+		{
+			ArenaRequestHandler = arenaRequestHandler;
+		}
+
+		protected override async Task ActionAsync(CancellationToken cancellationToken)
+		{
+			var responseRoundStates = (await ArenaRequestHandler.GetStatusAsync(cancellationToken).ConfigureAwait(false))
+				.ToDictionary(round => round.Id);
+
+			var updatedRoundStates = responseRoundStates.Where(round => RoundStates.ContainsKey(round.Key));
+			var newRoundStates = responseRoundStates.Where(round => !RoundStates.ContainsKey(round.Key));
+			var removedRoundStates = RoundStates.Where(round => !responseRoundStates.ContainsKey(round.Key));
+
+			List<uint256> roundsToUpdate = new();
+
+			foreach (var updatedRound in updatedRoundStates.Select(round => round.Value))
+			{
+				var oldRound = RoundStates[updatedRound.Id];
+				if (oldRound != updatedRound)
+				{
+					roundsToUpdate.Add(updatedRound.Id);
+				}
+			}
+
+			foreach (var round in newRoundStates.Select(round => round.Value))
+			{
+				roundsToUpdate.Add(round.Id);
+			}
+
+			foreach (var round in removedRoundStates.Select(round => round.Value))
+			{
+				// TODO: set all associated tasks to failed.
+			}
+
+			RoundStates = updatedRoundStates.Union(newRoundStates).ToImmutableDictionary();
+
+			ExecuteAwaiters(roundsToUpdate, RoundStates);
```

Comment:
I think that doing this could simplify the `ExcecuteAwaiters` method.

```suggestion
          var roundsToExecute = RoundStates.Where(r => roundsToUpdate.Contains(r.Key));
          ExecuteAwaiters(roundsToExecute);
```